src/journal.rs

```diff
@@ -0,0 +1,181 @@
+            line.push_str(&format!("{}", call));
+
+            line.push_str(&format!("{}", color::Fg(color::White)));
+            line.push_str(" |");
+
+            should_print = true;
+          }
+          JournalOp::FnCall(fn_call) => {
+            line.push_str(&format!("{}", color::Fg(color::Yellow)));
+            line.push_str(&format!("{}", fn_call));
+
+            line.push_str(&format!("{}", color::Fg(color::White)));
+            line.push_str(" |");
+
+            should_print = true;
+          }
+          JournalOp::Push(push) => {
+            line.push_str(&format!("{}", color::Fg(color::Green)));
+            line.push_str(&format!("{}", push));
+
+            should_print = true;
+          }
+          JournalOp::Pop(pop) => {
+            line.push_str(&format!("{}", color::Fg(color::Red)));
+            line.push_str(&format!("{}", pop));
+
+            should_print = true;
+          }
+          _ => {}
+        }
+      }
+
+      if should_print {
+        line.push_str(&format!("{}", color::Fg(color::Reset)));
+
+        let bullet_symbol = match entry.scoped {
+          true => format!("{}*", "  ".repeat(entry.scope)),
+          false => {
+            format!("{}!", "  ".repeat(entry.scope))
+          }
+        };
+        write!(f, " {} ", bullet_symbol)?;
+
+        if i != len - 1 {
+          writeln!(f, "{}", line)?;
+        } else {
+          write!(f, "{}", line)?;
+        }
+      }
+    }
+
+    Ok(())
+  }
+}
+
+impl Journal {
+  pub fn new() -> Self {
+    Self::default()
+  }
+
+  pub fn op(&mut self, op: JournalOp) {
+    if matches!(op, JournalOp::Commit) {
+      return self.commit();
+    }
+
+    self.current.push(op);
+  }
+
+  pub fn commit(&mut self) {
+    if !self.current.is_empty() {
+      self.ops.extend(mem::take(&mut self.current));
+      self.commits.push(self.ops.len());
+      self.ops.push(JournalOp::Commit);
+    }
+  }
+
+  pub fn clear(&mut self) {
+    self.ops.clear();
+    self.current.clear();
+    self.commits.clear();
+  }
+
+  fn entries(&self, max_commits: usize) -> Vec<JournalEntry> {
+    let mut entries: Vec<JournalEntry> = Vec::new();
+
+    let start = match max_commits >= self.commits.len() {
+      false => self
+        .commits
+        .get(self.commits.len() - max_commits - 1)
+        .map(|index| index + 1)
+        .unwrap_or(0),
+      true => 0,
+    };
+
+    let mut scope = 0;
+    let mut ops: Vec<JournalOp> = Vec::new();
+    let mut scoped: Vec<bool> = vec![true];
+    for op in self.ops.iter().skip(start) {
+      match op {
+        JournalOp::Commit => {
+          entries.push(JournalEntry::new(
+            ops,
+            scope,
+            *scoped.last().unwrap_or(&true),
+          ));
+          ops = Vec::new();
+        }
+        JournalOp::FnStart(is_scoped) => {
+          scope += 1;
+          scoped.push(*is_scoped);
+        }
+        JournalOp::FnEnd => {
+          scope = scope.saturating_sub(1);
```

Comment:
Using saturating sub due to a bug where it would go below zero. It shouldn't underflowed, but I didn't know why it was, so I handled it. After implementing this and running the Stack code that caused the subtraction error, nothing looked out of place in the journal, so I think it's fine if we ignore that case and simply saturate the subtraction.

Reply:
You could try using `wrapping_add` to see what the output looks like when it overflows.